src/common/btree.c

```diff
@@ -1257,60 +1334,123 @@ btr_probe(struct btr_context *tcx, dbtree_probe_opc_t probe_opc,
 		D_ASSERTF(cmp == BTR_CMP_EQ, "Hash collision is unsupported\n");
 	}
 
+	vbt.nd_mmid = nd_mmid;
+	vbt.at = at;
+	vbt.intent = intent;
+
+again:
 	switch (probe_opc & ~BTR_PROBE_MATCHED) {
 	default:
 		D_ASSERT(0);
 	case BTR_PROBE_FIRST:
+		do {
+			vbt.nd_mmid = tcx->tc_trace[level].tr_node;
+			vbt.at = tcx->tc_trace[level].tr_at;
+			rc = btr_check_visibility(tcx, &vbt);
+		} while (rc == 0 && btr_probe_next(tcx));
+
+		if (rc <= 0)
+			rc = PROBE_RC_NONE;
+		goto out;
+
 	case BTR_PROBE_LAST:
-		rc = PROBE_RC_OK;
+		do {
+			vbt.nd_mmid = tcx->tc_trace[level].tr_node;
+			vbt.at = tcx->tc_trace[level].tr_at;
+			rc = btr_check_visibility(tcx, &vbt);
+		} while (rc == 0 && btr_probe_prev(tcx));
+
+		if (rc <= 0)
+			rc = PROBE_RC_NONE;
 		goto out;
 
 	case BTR_PROBE_EQ:
 		if (cmp == BTR_CMP_EQ) {
-			rc = PROBE_RC_OK;
-			goto out;
+			rc = btr_check_visibility(tcx, &vbt);
+			if (rc > 0)
+				goto out;
+
+			/* Current pos is invisible, it can be used for the
+			 * follow-on insert if applicable.
+			 */
```

Comment:
I wonder if the concept of visibility shouldn't be tracked by the vos tree rather than internally in the btree.

Reply:
Currently, there are some users outside of VOS needs to make decision based on related record's visibility. For example, we want to reuse former aborted slot in the b+tree when insert new record. The insert logic is controlled inside common/btree.c, VOS only offers the to_rec_alloc/to_rec_free API. So we need trigger the visibility check outside of VOS.

Reply:
I agree with Jeff that the visibility check should be done by vos tree (when get leaf on fetch, iteration or preparing next level tree root) but not common btree.

To support 'reusing' aborted record, we may introduce another interface to check if the leaf node is an aborted one which could be cleared, and this interface should only be called on CMP_EQ case, so that once eligible aborted record is detected on insert, we can clear it immediately. I suppose that requires less changes in common btree, also it won't expose the concept of 'visibility' to common btree (which looks kind of layer violation to me).

Reply:
After discussion with Fanyong, I tend to think that it's ok to put the 'visibility' check in btree/evtree.

Reply:
so we will do visibility check for all matched cases, then we should do some code cleanup and do the check in one place, and re-probe if necessary. 

I tried to re-org the code but it took me a lot of time and ended nowhere, anyhow, we really should try this in the future.